### ISSUE TITLE
"location" command clarification

### ISSUE THREAD

**Issue by luorax** [premake/premake-core#564]
So today I came back to the premake5 repository to see how development goes, because I just love this project & build system. I quickly noticed that the first official tutorial is up, so I thought I'd check it out. Then I thought I found a bug in it - it only had a workspace level "location" call, yet it claimed that all generated files would be put into the "Generated" folder. Quoting the wiki page for [location](https://github.com/premake/premake-core/wiki/location):

> Note that unlike other values, location does not automatically propagate to the contained projects. Projects will use their default location unless explicitly overridden.

So my question is: is the wiki page for location deprecated/wrong, or is it a bug? Either way, one or the other should be fixed. I suspect it's the wiki page, but I don't want to go ahead and modify it without knowing for sure that this is the intended behavior.

**Comment by tvandijck** [comment]
Well, that is certainly not the behavior I'm seeing indeed... and also not what the code does:

this is in the oven.project.bake()
```lua
self.location = self.location or wks.location or self.basedir
```

So it is true that if the project.location is set it override it, and if the workspace.location is not set it will have some kind of default... but the documentation is incorrect on it not propagating.

**Comment by samsinsane** [comment]
Related, the documentation also says:
> By default, workspace and project files are generated into the same directory as the script that defines them.

Which is broken because the value propagates and `wks.location` is defined by:
```lua
self.location = self.location or self.basedir
```
And `self.basedir` is defined by:
```lua
self.basedir = os.getcwd()
```

I went back through the history and I can't see where the described functionality would ever work. `self.basedir` was defined as `os.getcwd()` as far back as 2009, I'm sure something, somewhere, changed along the way that has caused this to start happening. Any ideas on how we can achieve the documented functionality and *not* have `self.location` be `nil`? It looks like a bit of code relies on the various locations to be defined.

**Comment by starkos** [comment]
> I went back through the history and I can't see where the described functionality would ever work. self.basedir was defined as os.getcwd() as far back as 2009, I'm sure something, somewhere, changed along the way that has caused this to start happening.

It works because Premake's implementation of luaL_loadfile() changes to the directory containing the script before running it. So when those snippets of code are executed, the current directory is in fact the one containing the script.

That said, we now have the `_SCRIPT_DIR` global, which would be a more reliable source for that path.

**Comment by samsinsane** [comment]
> It works because Premake's implementation of luaL_loadfile() changes to the directory containing the script before running it. So when those snippets of code are executed, the current directory is in fact the one containing the script.

That's not quite what I meant, the code that has been posted above has never changed from what I could tell when I went through the history of the repo. So how does `self.location = self.location or wks.location or self.basedir` ever set `self.location` to `self.basedir`? `wks.location` is always set.

Just to clarify, do you experience this bug too? I think I've always had this bug :(

**Comment by samsinsane** [comment]
This has been fixed by #733
